var/spack/repos/builtin/packages/py-jupyterhub/package.py

```diff
@@ -10,12 +10,11 @@ class PyJupyterhub(PythonPackage):
     """Multi-user server for Jupyter notebooks."""
 
     homepage = "https://pypi.org/project/jupyterhub"
-    url      = "https://pypi.io/packages/source/j/jupyterhub/jupyterhub-0.9.4.tar.gz"
+    url      = "https://pypi.io/packages/source/j/jupyterhub/jupyterhub-1.0.0.tar.gz"
 
-    version('0.9.4',    sha256='86b1cce446d4e8347e26913878858fc8964d103fde19b606fe37ccc5188d629d')
+    version('1.0.0',    sha256='33541a515a041b9a518ca057c1c4ab4215a7450fdddc206401713ee8137fa67f')
 
     depends_on('python@3.5:')
-    depends_on('node-js', type=('build', 'run'))
     depends_on('py-setuptools', type='build')
     depends_on('py-python-dateutil', type='run')
```

Comment:
All of these other dependencies should be `type=('build', 'run')`. Also, can you add the version constraints:
```
alembic
async_generator>=1.8
certipy>=0.1.2
entrypoints
jinja2
oauthlib>=3.0
pamela
prometheus_client>=0.0.21
python-dateutil
requests
SQLAlchemy>=1.1
tornado>=5.0
traitlets>=4.3.2
```

Reply:
There is no hash for traitlets 4.3.2. Is that the right constraint to use?

Reply:
@wspear you can do something like this:
```python
depends_on('py-traitlets@4.3.2:', type=('build', 'run'))
```
The trailing colon tells Spack that it requires version 4.3.2 *or greater*. In the case of `py-traitlets`, it looks like you'll need to add a newer version to that package in order to build this package. It doesn't need to be exactly 4.3.2, but at least 4.3.2.

Reply:
I am already using that syntax. The issue is the absence of explicit higher versions with hashes in the dependent packages. Are you sure about these build requirements? It was building and installing for me before I added them (but I hadn't done any run tests)). Tornado 5 doesn't exist either.

depends_on('py-tornado@5.0:', type=('build', 'run'))
...
`==> Installing py-tornado
==> Searching for binary cache of py-tornado
==> No binary for py-tornado found: installing from source
==> Warning: There is no checksum on file to fetch py-tornado@5.0 safely.
==>   Fetch anyway? [y/N] y
==> Fetching with no checksum.
  Add a checksum or use --no-checksum to skip this check.
==> Fetching http://oaciss.uoregon.edu/e4s/x86_64/py-tornado/py-tornado-5.0.tar.gz

curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from http://oaciss.uoregon.edu/e4s/x86_64/py-tornado/py-tornado-5.0.tar.gz failed.
==> Fetching https://github.com/tornadoweb/tornado/archive/v5.0.tar.gz
#=#=-#    #                                                                                                                        
curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from https://github.com/tornadoweb/tornado/archive/v5.0.tar.gz failed.
==> Error: FetchError: All fetchers failed for py-tornado-5.0-7cy4v2cmrqbestan5xszqyrh52fgqv7t

/home/wspear/bin/spack/lib/spack/spack/package.py:1064, in do_fetch:
       1061                raise FetchError("Will not fetch %s" %
       1062                                 self.spec.format('{name}{@version}'), ck_msg)
       1063
  >>   1064        self.stage.create()
       1065        self.stage.fetch(mirror_only)
       1066        self._fetch_time = time.time() - start_time
       1067

`

Reply:
> The issue is the absence of explicit higher versions with hashes in the dependent packages.

Yes, you need to add newer versions to those packages.

> Are you sure about these build requirements?

Yep, they come directly from `install_requires` in `setup.py`. See https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#python-dependencies

> It was building and installing for me before I added them

That's because setuptools manually downloads and installs the dependencies for you if it can't find them. This breaks reproducibility, and means the package can't be installed at all on air-gapped networks, so it's important to get the dependency constraints right in the package.

Reply:
This request is now dependent on #13080, #13079, #13078, #13077, #13068 and #13075. I'll wait until I can test it with those merged before I make any changes.

Reply:
@wspear looks like all of these dependency PRs are now merged!